var/spack/repos/builtin/packages/py-numpy/package.py

```diff
@@ -67,68 +70,93 @@ def setup_dependent_package(self, module, dep_spec):
-                if '^openblas' in spec:
-                    f.write('[openblas]\n')
-                    f.write('libraries=%s\n'    % names)
-                elif '^mkl' in spec:
-                    # numpy does not expect system libraries needed for MKL
-                    # here.
-                    # names = [x for x in names if x.startswith('mkl')]
-                    # FIXME: as of @1.11.2, numpy does not work with separately
-                    # specified threading and interface layers. A workaround is
-                    # a terribly bad idea to use mkl_rt. In this case Spack
-                    # won't no longer be able to guarantee that one and the
-                    # same variant of Blas/Lapack (32/64bit, threaded/serial)
-                    # is used within the DAG. This may lead to a lot of
-                    # hard-to-debug segmentation faults on user's side. Users
-                    # may also break working installation by (unconciously)
-                    # setting environment variable to switch between different
-                    # interface and threading layers dynamically. From this
-                    # perspective it is no different from throwing away RPATH's
-                    # and using LD_LIBRARY_PATH throughout Spack.
-                    f.write('[mkl]\n')
-                    f.write('mkl_libs=%s\n'     % 'mkl_rt')
-                elif '^atlas' in spec:
-                    f.write('[atlas]\n')
-                    f.write('atlas_libs=%s\n'   % names)
+            lapackblas_includes.add(spec['blas'].prefix.include)
+            lapackblas_libraries += spec['blas'].blas_libs
+
+        # Each BLAS/LAPACK requires a custom site.cfg
+
+        # Generic build notes (very out-of-date):
+        # http://www.scipy.org/scipylib/building/linux.html
+
+        # Example site.cfg:
+        # https://github.com/numpy/numpy/blob/master/site.cfg.example
+        with open('site.cfg', 'w') as f:
+            libraries = ','.join(lapackblas_libraries.names)
+
+            if '^openblas' in spec:
+                f.write('[openblas]\n')
+                f.write('libraries = {0}\n'.format(libraries))
+
+            elif '^mkl' in spec:
+                # Intel MKL build notes:
+                # https://software.intel.com/en-us/articles/numpyscipy-with-intel-mkl
+
+                # library_dirs are in compilers_and_libraries_2017/linux/mkl/lib/intel64  # noqa
+                # include_dirs are in compilers_and_libraries_2017/linux/mkl/include      # noqa
+                lapackblas_includes = set()
+                for libdir in lapackblas_libraries.directories:
+                    incdir = os.path.dirname(os.path.dirname(libdir))
+                    incdir = join_path(incdir, 'include')
+                    lapackblas_includes.add(incdir)
+
+                f.write('[mkl]\n')
+                f.write('mkl_libs = {0}\n'.format('mkl_rt'))
```

Comment:
the point is that `mkl_rt` completely ingores any variants we use in `mkl` such as 32bit or 64bit, single threaded or multithreaded. So you can think of `mkl_rt` as an easy-to-make-it-work approach, but in Spack it's something we decided not to use to be rigorous about which version of Blas/Lapack is used within the DAG. So yes, it's hacky from Spack's perspective. Most likely the reasons why Intel recommends it is because `numpy` does NOT work otherwise, although i should!